sdk/src/main/scala/io/horizen/account/mempool/MempoolMap.scala

```diff
@@ -334,34 +337,31 @@ class MempoolMap(
     val listOfRejectedBlocksTxs = rejectedBlocks.flatMap(_.transactions)
     val rejectedTransactionsByAccount = listOfRejectedBlocksTxs.groupBy(_.getFrom)
 
-    rejectedTransactionsByAccount.foreach { case (account, rejectedTxs) =>
+    var newExecTxs = rejectedTransactionsByAccount.flatMap { case (account, rejectedTxs) =>
       val latestNonceAfterAppliedTxs = appliedTxNoncesByAccount.remove(account)
-      if (latestNonceAfterAppliedTxs.isDefined)
-        updateAccount(account, latestNonceAfterAppliedTxs.get, rejectedTxs)
-      else
-        readdedTxs = readdedTxs ++ updateAccountWithRevertedNonce(account, rejectedTxs)
+      val promotedTxs = latestNonceAfterAppliedTxs match {
+        case Some(latestNonce) => updateAccount(account, latestNonce, rejectedTxs)
+        case None => updateAccountWithRevertedNonce(account, rejectedTxs)
+      }
+      promotedTxs
     }
-    appliedTxNoncesByAccount.foreach { case (account, nonce) =>
+
+    newExecTxs = newExecTxs ++ appliedTxNoncesByAccount.flatMap { case (account, nonce) =>
       updateAccount(account, nonce)
     }
 
     removeTimedoutTransactions()
 
     checkMempoolSize()
 
-    if (readdedTxs.size > this.size)
-      // If the number of the transactions to be readded exceed the mempool size, we just return the mempool
-      // because the mempool now contains only transactions taken from the removed blocks because are newer than
-      // the transactions previously contained
-      txCache.values.toSeq
-    else
-      readdedTxs
+    //newExecTxs is not changed after checkMempoolSize because the event is sent even if the corresponding tx is evicted
+    newExecTxs
```

Comment:
So, if transaction reentered the meempool from the reverted block or was promoted (was present before as non-executable) we will return it in any case?
But what if it was then removed from the mempool?

Reply:
In theory, txs evicted are are like any other valid txs added to the mempool and later evicted because the mempool is too big, there are no events for this. In this case this is almost at the same time, but it is consistent

Reply:
ok